worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,142 @@
+
+package controlsocket
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/state"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this worker is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	userCreator = "admin"
+)
+
+func (w *Worker) registerHandlers(handle func(pattern string, handler http.Handler)) {
+	// TODO: does this API structure make sense?
+	handle("/metrics-users/add", http.HandlerFunc(w.handleAddMetricsUser))
+	handle("/metrics-users/remove", http.HandlerFunc(w.handleRemoveMetricsUser))
+}
+
+func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	queryParams := req.URL.Query()
+
+	if !queryParams.Has("username") {
+		w.writeResponse(resp, http.StatusBadRequest, `missing "username" query parameter`)
+		return
+	}
+	username := queryParams.Get("username")
+
+	if !queryParams.Has("password") {
+		w.writeResponse(resp, http.StatusBadRequest, `missing "password" query parameter`)
+		return
+	}
+	password := queryParams.Get("password")
+
+	err := validateMetricsUsername(username)
+	if err != nil {
+		w.writeResponse(resp, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	if password == "" {
+		w.writeResponse(resp, http.StatusBadRequest, "empty password")
+		return
+	}
+
+	user, err := w.state.AddUser(username, username, password, userCreator)
+	if err != nil {
+		w.writeResponse(resp, http.StatusInternalServerError, "failed to create user %q: %v", username, err)
+		return
+	}
+	userTag := user.UserTag()
+	// We should clean up the user if the later grant access fails.
+	cleanup := func() {
+		err := w.state.RemoveUser(userTag)
+		if err != nil {
+			// Best we can do here is log an error.
+			w.logger.Warningf("add metrics user failed, but could not clean up user %q: %v",
+				username, err)
+		}
+	}
+
+	model, err := w.state.Model()
+	if err != nil {
+		w.writeResponse(resp, http.StatusInternalServerError, "retrieving current model: %v", err)
+		cleanup()
+		return
+	}
+
+	_, err = model.AddUser(state.UserAccessSpec{
+		User:      userTag,
+		CreatedBy: names.NewUserTag(userCreator),
+		Access:    permission.ReadAccess,
+	})
+	if err != nil {
+		w.writeResponse(resp, http.StatusInternalServerError, "adding user %q to model %q: %v",
+			username, bootstrap.ControllerModelName, err)
+		return
+	}
+
+	w.writeResponse(resp, http.StatusOK, "successfully added metrics user %q", username)
+}
+
+func (w *Worker) handleRemoveMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	queryParams := req.URL.Query()
+
+	if !queryParams.Has("username") {
+		w.writeResponse(resp, http.StatusBadRequest, `missing "username" query parameter`)
+		return
+	}
+	username := queryParams.Get("username")
```

Comment:
Is there no way to hold some sort of uuid (nonce) with this, it feels dangerous to be able to rely on a username here. (creates a user, deletes a user, creates a user with the same name behind your back).

Reply:
All the state methods only accept a username. I think it's assumed all throughout state that the username is the nonce, and it uniquely identifies the user. So I don't think there's much we can do here.

Reply:
If it alleviates your concern slightly, these usernames are all prefixed with `juju-metrics-`, and no-one should be messing around with that "namespace" except this worker.